Added Label Alignment

T-rav opened the pull request:
I have added a feature to allow the user to specify how labels on edges behave.
This new behavior aligns the label to the line, either above or on it.
There is a new configuration property labelAlignment for edges. 
It has two values : line-center or line-above.
If no value is given the current behavior is used.

I have added a working example 35_label_alignment.html showing all three cases of labels on edges.

AlexDM0 commented:
Hi Trav,

Could you add documentation for this?

Regards,

Alex

AlexDM0 commented:
I'm not sure I can tell what exactly is new in the example you provided.. 
http://htmlpreview.github.io/?https://github.com/T-rav/vis/blob/master/examples/network/35_label_alignment.html

The orientation of the labels is still absolute horizontal

T-rav commented:
Looks like a merge issue. I do not see my work. Will check it out.
On 14 Jan 2015 11:40, "Alex" <notifications@github.com> wrote:

> I'm not sure I can tell what exactly is new in the example you provided..
>
> http://htmlpreview.github.io/?https://github.com/T-rav/vis/blob/master/examples/network/35_label_alignment.html
>
> The orientation of the labels is still absolute horizontal
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/almende/vis/pull/566#issuecomment-69891228>.
>

AlexDM0 commented:
Hi T-rav,

Could you also simplify the example? Since this is about labels, we do not need the (rounded) images. We have another example for that. Just the same node setup as the very basic example 1, but with edge labels in all alignments would be perfect.

Regards,

Alex

T-rav commented:
Hi Alex,

I have made the example simpler as you requested.
I have also committed the dist/vs.js and dist/vs.min.js files. 
The example was not working due to the fact these where not committed.

Cheers,
Trav

AlexDM0 commented:
Hi Trav,

Could you merge it with the latest develop? There have been changes to the code you're using lately so I'd rather you merge with that beforehand. Could you add documentation? Finally, it seems if you're offering line-center and line-above, that line-below would be a natural addition. Is that possible?

Regards,

Alex

T-rav commented:
I will merge with the latest develop.
And then add a line-below option as well. 
As for documentation, where do I need to add it?

AlexDM0 commented:
Hi Trav,

There's an HTML file in /docs/network.html. Those are the docs we use on the website. If you could add the option there in the appropriate place that would be great! 

Your picture seems to be orphaned now that you changed the example. Could you remove it from the merge? We have had some problems with merging when the dist folder was also in the merge. The problem is that if we merge manually, git does not always pick up on that the pull req was performed. I'd like it to show as accepted if possible, to your credit. If the dist folder is not present, everything is a bit easier :)

Thank you very much for the contribution!

Regards,

Alex

T-rav commented:
Hi Alex,

I have merged from your develop branch.
I have removed the picture of me.
I have also added a line-below option and included it in the example.
And I have updated the network.html docs to reflect this option in Edge configuration.

Cheers,
Trav

AlexDM0 commented on lib/network/Edge.js:
```diff
@@ -580,33 +583,122 @@ Edge.prototype._label = function (ctx, text, x, y) {
+	
+	ctx.save();
+	
+<<<<<<< HEAD
```
this has not merged correctly it seems...

AlexDM0 commented on lib/network/Edge.js:
```diff
@@ -580,33 +583,122 @@ Edge.prototype._label = function (ctx, text, x, y) {
+	if (this.options.fontFill !== undefined && this.options.fontFill !== null && this.options.fontFill !== "none") {
+		ctx.fillStyle = this.options.fontFill;
+		
+<<<<<<< HEAD
```
merge issues

AlexDM0 commented:
Hi Trav,

Sorry to make this so complicated.. It seems something is wrong with the merge you performed. I added some notes to it. Could you check it?

Regards,

Alex

AlexDM0 commented on lib/network/Edge.js:
```diff
@@ -580,33 +583,122 @@ Edge.prototype._label = function (ctx, text, x, y) {
+	ctx.textAlign = "center";
+	if(labelAlignment == 'line-above'){
+		ctx.textBaseline =  "bottom";
+<<<<<<< HEAD
```
additional merge issues

T-rav commented:
Should be fixed now. Sorry about that.

AlexDM0 commented:
Hi Travis,

It does not seem possible to use this with the smoothCurves which we have enabled by default. I'll attempt a merge and see what we get :).

Regards,

Alex

T-rav commented:
I am a bit confused about the smoothCurves issue....
If smoothCurves is on by default it should have caused issues when I developed the feature.
When moving a node edge will bend and curve but the text alignment follows a sensible flow.
Let me know if you need anything else done after the merge.

AlexDM0 commented:
Hi Travis,

It seems to work, the smoothCurves seem to sort themselves out. My bad. I'm tweaking it a little to be more stable across browsers. Specifically: http://www.w3schools.com/tags/canvas_textbaseline.asp

the different browsers leave more open to interpretation using top and bottom, so I prefer alphabetic and hanging.

I've made the default entree be 'horizontal' as instead of undefined. I'd prefer all options to be defined for clarity.

It seems the labelStrokes have been broken. I'll fix them.

I have merged everything is it's working. Thank you for your contribution!

Regards,

Alex

T-rav commented:
Thanks for the patch ups and merging the feature in.
